core/valor_core/utilities.py

```diff
@@ -0,0 +1,945 @@
+        raise ValueError("Raster is empty, cannot create bounding box.")
+
+    ymin, ymax = np.where(rows)[0][[0, -1]]
+    xmin, xmax = np.where(cols)[0][[0, -1]]
+
+    return schemas.Box.from_extrema(xmin, xmax + 1, ymin, ymax + 1)
+
+
+def _convert_raster_to_polygon(raster: np.ndarray) -> schemas.Polygon:
+    """Convert a raster mask to a Polygon."""
+    if raster.ndim != 2:
+        raise ValueError("Raster must be a 2D array.")
+
+    mask = (raster > 0).astype(np.uint8)
+    rows, cols = np.where(mask > 0)
+
+    if len(rows) == 0 or len(cols) == 0:
+        raise ValueError("Raster is empty, cannot create a polygon.")
+
+    contours = []
+    for r, c in zip(rows, cols):
+        if (
+            (r > 0 and mask[r - 1, c] == 0)
+            or (r < mask.shape[0] - 1 and mask[r + 1, c] == 0)
+            or (c > 0 and mask[r, c - 1] == 0)
+            or (c < mask.shape[1] - 1 and mask[r, c + 1] == 0)
+        ):
+            contours.append((c, r))
+
+    if not contours:
+        raise ValueError("No contours found in raster.")
+
+    contours = sorted(contours, key=lambda p: (p[1], p[0]))
+
+    polygon = [[(x, y) for x, y in contours] + [contours[0]]]
+
+    return schemas.Polygon.from_dict(
+        {"type": "Polygon", "coordinates": polygon}
+    )
+
+
+def _convert_polygon_to_box(polygon: schemas.Polygon) -> schemas.Box:
+    """Convert a Polygon to a Box."""
+
+    boundary = polygon.boundary
+
+    xmin = min(point[0] for point in boundary)
+    xmax = max(point[0] for point in boundary)
+    ymin = min(point[1] for point in boundary)
+    ymax = max(point[1] for point in boundary)
+
+    return schemas.Box.from_extrema(xmin, xmax, ymin, ymax)
+
+
+def _identify_most_detailed_annotation_type(
+    df: pd.DataFrame,
+) -> enums.AnnotationType:
+    """
+    Identify the most detailed annotation type present in the DataFrame.
+
+    The function checks the columns in the given DataFrame for non-null values
+    to determine the most detailed annotation type in the following order:
+    raster, polygon, bounding_box. If none of these types are present,
+    it returns AnnotationType.NONE.
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        DataFrame containing the annotations.
+
+    Returns
+    -------
+    enums.AnnotationType
+        The most detailed annotation type present in the DataFrame.
+    """
+
+    if df["raster"].notnull().any():
+        return enums.AnnotationType.RASTER
+
+    elif df["polygon"].notnull().any():
+        return enums.AnnotationType.POLYGON
+
+    elif df["bounding_box"].notnull().any():
+        return enums.AnnotationType.BOX
+
+    else:
+        return enums.AnnotationType.NONE
+
+
+def _add_converted_geometry_column(
+    df: pd.DataFrame,
+    target_type: enums.AnnotationType,
+) -> pd.DataFrame:
+    """
+    Add a column with converted geometries to the DataFrame.
+
+    The function checks that each annotation contains only one type of geometry
```

Comment:
Why are we regressing on functionality? An annotation can include multiple representations of itself in the current backend.

YOLO inferences for example provides both bitmasks and bounding boxes per annotation.

Reply:
there are no integration or functional tests where this is true, so this is part of the "blind risk" mentioned in "Issues to Fix..". I can consider relaxing this assumption when implementing better tests in the next PR